util/pkg/reflectutils/access_test.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reflectutils
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+type fakeObject struct {
+	Spec   fakeObjectSpec
+	Status fakeObjectStatus
+}
+
+func (o *fakeObject) String() string {
+	b, err := json.Marshal(o)
+	if err != nil {
+		return fmt.Sprintf("(error:%v)", err)
+	}
+	return string(b)
+}
+
+type fakeObjectSpec struct {
+	Containers []fakeObjectContainers
+}
+
+type fakeObjectContainers struct {
+	Image     string
+	Policy    *fakeObjectPolicy
+	Resources *fakeObjectResources
+}
+
+type fakeObjectPolicy struct {
+	Name  string
+	Allow bool
+}
+
+type fakeObjectResources struct {
+	Limits   map[string]int
+	Requests map[string]int
+}
+
+type fakeObjectStatus struct {
+}
+
+func toJSON(s string) []byte {
+	s = strings.ReplaceAll(s, "'", "\"")
+	return []byte(s)
+}
+
+func TestSet(t *testing.T) {
+	grid := []struct {
+		Name     string
+		Input    string
+		Expected string
+		Path     string
+		Value    string
+	}{
+		{
+			Name:     "simple setting",
+			Input:    "{ 'spec': { 'containers': [ {} ] } }",
+			Expected: "{ 'spec': { 'containers': [ { 'image': 'hello-world' } ] } }",
+			Path:     "Spec.Containers[0].Image",
+			Value:    "hello-world",
+		},
+		{
+			Name:     "setting with wildcard",
+			Input:    "{ 'spec': { 'containers': [ {} ] } }",
+			Expected: "{ 'spec': { 'containers': [ { 'image': 'hello-world' } ] } }",
+			Path:     "Spec.Containers[*].Image",
+			Value:    "hello-world",
+		},
+		{
+			Name:     "creating missing objects",
+			Input:    "{ 'spec': { 'containers': [ {} ] } }",
+			Expected: "{ 'spec': { 'containers': [ { 'policy': { 'name': 'allowed' } } ] } }",
+			Path:     "Spec.Containers[0].Policy.Name",
+			Value:    "allowed",
+		},
+		// Not sure if we should do this...
```

Comment:
I think if it's effectively appending an element, then it's okay.

Reply:
We insert intermediate elements when they're nil pointers, which seems analogous. If it's not valid after processing all setters then API validation will catch that.